src/reader/npz_reader.py

```diff
@@ -40,48 +41,81 @@ def read(self, epoch_number):
-                                        int(total_samples_per_rank * (self.my_rank + 1) / self.batch_size))
-                num_sets = list(range(part_start, part_end))
-            total += len(num_sets)
+                part_start, part_end = (int(total_samples_per_rank * self.my_rank),
+                                        int(total_samples_per_rank * (self.my_rank + 1)))
+                sample_index_list = list(range(part_start, part_end))
+
             if self.sample_shuffle != Shuffle.OFF:
                 if self.sample_shuffle == Shuffle.SEED:
                     random.seed(self.seed)
-                random.shuffle(num_sets)
-            for num_set in num_sets:
-                # Should we check if profiling is enabled and if we're using PT?
-                with tf.profiler.experimental.Trace('HDF5 Input', step_num=num_set / self.batch_size, _r=1):
-                    progress(count, total, "Reading NPZ Data")
-                    count += 1
-                    images = element['dataset'][:][:][num_set * self.batch_size:(num_set + 1) * self.batch_size - 1]
-                yield images
+                random.shuffle(sample_index_list)
+            for sample_index in sample_index_list:
+                count += 1
+                logging.info(f"{utcnow()} num_set {sample_index} current batch_size {len(batch)}")
+                my_image = self._dataset[index]['data'][..., sample_index]
+                logging.debug(f"{utcnow()} shape of image {my_image.shape} self.max_dimension {self.max_dimension}")
+
+                my_image_resized = np.resize(my_image, (self.max_dimension, self.max_dimension))
+                logging.debug(f"{utcnow()} new shape of image {my_image_resized.shape}")
+                batch.append(my_image_resized)
+                is_last = 0 if count < total else 1
+                if is_last:
+                    while len(batch) is not self.batch_size:
+                        batch.append(np.random.rand(self.max_dimension, self.max_dimension))
+                if len(batch) == self.batch_size:
+                    batch = np.array(batch)
+                    yield is_last, batch
+                    batch = []
+
+    def read_index(self, index):
+        file_index = math.floor(index / self.num_samples)
+        if self._dataset[file_index]["data"] is None:
+            with np.load(self._dataset[file_index]["file"], allow_pickle=True) as data:
+                self._dataset[file_index]['data'] = data["x"]
+        element_index = index % self.num_samples
+        my_image = self._dataset[file_index]['data'][..., element_index]
+        logging.info(f"{utcnow()} shape of image {my_image.shape} self.max_dimension {self.max_dimension}")
+        my_image_resized = np.resize(my_image, (self.max_dimension, self.max_dimension))
+        logging.info(f"{utcnow()} new shape of image {my_image_resized.shape}")
+        return my_image_resized
```

Comment:
In this case again, on the second epoch, data will be read from the memory. 

It is not truly the on demand way of doing things. 

I think the key here is because you are using ```self._dataset[file_index]['data']``` to hold all the data

Reply:
I can force the re-read in case of ON-DEMAND.

Reply:
Fixed